cloud/scope/machine.go

```diff
@@ -93,7 +95,16 @@ func (m *MachineScope) Cloud() cloud.Cloud {
 // Zone returns the FailureDomain for the GCPMachine.
 func (m *MachineScope) Zone() string {
 	if m.Machine.Spec.FailureDomain == nil {
-		return ""
+		region, err := m.ClusterGetter.Cloud().Regions().Get(context.TODO(), meta.GlobalKey(m.ClusterGetter.Region()))
+		if err != nil {
+			return ""
+		}
+
+		zones, err := m.ClusterGetter.Cloud().Zones().List(context.TODO(), filter.Regexp("region", region.SelfLink))
```

Comment:
region and zones info are already available in [gcpcluster_controller.go](https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/main/controllers/gcpcluster_controller.go#L169-L194). Could you find a way to reuse the information? 

Asking because we want to minimize calls to GCP API. There is a rate limit and this code will add 2 API calls per each instance creation.

Reply:
Yes, I had a doubt about this too. I did another approach of returning the zone in the [reconcile](https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/main/controllers/gcpcluster_controller.go#L169-L194) function

Reply:
> Yes, I had a doubt about this too. I did another approach of returning the zone in the [reconcile](https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/main/controllers/gcpcluster_controller.go#L169-L194) function

But that will cause `reconcile` function call again and API call will happen. I guess better is to store the zone list in some cluster state struct or variable.

One approach that I was thinking but not sure if it is valid or not that [ClusterScope FailureDomains](https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/421b2395a8c02e0f73be84e91e283193c3e718fd/cloud/scope/cluster.go#L136) function is returning the failure domains for the control plane cluster so can we take those failure domain and set one from them?

```go
if m.Machine.Spec.FailureDomain == nil {
    fd := m.ClusterGetter.FailureDomains().GetIDs()
    return *fd[0]
}
```

Reply:
Using FailureDomain info from Cluster is a good idea and already a pattern used in [load balancer](https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/421b2395a8c02e0f73be84e91e283193c3e718fd/cloud/services/compute/loadbalancers/reconcile.go#L90).

Reply:
Ok I have pushed accordingly